cmd/cosign/cli/verify/verify_blob.go

```diff
@@ -161,252 +149,109 @@ func (c *VerifyBlobCmd) Exec(ctx context.Context, blobRef string) error {
 			return fmt.Errorf("bundle does not contain cert for verification, please provide public key")
 		}
-		// b.Cert can either be a certificate or public key
-		certBytes := []byte(b.Cert)
-		if isb64(certBytes) {
-			certBytes, _ = base64.StdEncoding.DecodeString(b.Cert)
-		}
-		cert, err = loadCertFromPEM(certBytes)
-		if err != nil {
-			// check if cert is actually a public key
-			co.SigVerifier, err = sigs.LoadPublicKeyRaw(certBytes, crypto.SHA256)
-		} else {
-			if c.CertChain == "" {
-				co.RootCerts, err = fulcio.GetRoots()
-				if err != nil {
-					return fmt.Errorf("getting Fulcio roots: %w", err)
-				}
-				co.IntermediateCerts, err = fulcio.GetIntermediates()
-				if err != nil {
-					return fmt.Errorf("getting Fulcio intermediates: %w", err)
-				}
-				co.SigVerifier, err = cosign.ValidateAndUnpackCert(cert, co)
-				if err != nil {
-					return fmt.Errorf("verifying certificate from bundle: %w", err)
-				}
-			} else {
-				// Verify certificate with chain
-				chain, err := loadCertChainFromFileOrURL(c.CertChain)
-				if err != nil {
-					return err
-				}
-				co.SigVerifier, err = cosign.ValidateAndUnpackCertWithChain(cert, chain, co)
+		// We have to condition on this because sign-blob may not output the signing
+		// key to the bundle when there is no tlog upload.
+		if b.Cert != "" {
+			// b.Cert can either be a certificate or public key
+			certBytes := []byte(b.Cert)
+			if isb64(certBytes) {
+				certBytes, _ = base64.StdEncoding.DecodeString(b.Cert)
+			}
+			cert, err = loadCertFromPEM(certBytes)
+			if err != nil {
+				// check if cert is actually a public key
+				co.SigVerifier, err = sigs.LoadPublicKeyRaw(certBytes, crypto.SHA256)
 				if err != nil {
-					return fmt.Errorf("verifying certificate from bundle with chain: %w", err)
+					return fmt.Errorf("loading verifier from bundle: %w", err)
 				}
 			}
 		}
+		opts = append(opts, static.WithBundle(b.Bundle))
+	}
+	// Set an SCT if provided via the CLI.
+	if c.SCTRef != "" {
+		sct, err := os.ReadFile(filepath.Clean(c.SCTRef))
 		if err != nil {
-			return fmt.Errorf("loading verifier from bundle: %w", err)
+			return fmt.Errorf("reading sct from file: %w", err)
 		}
-		bundle = b.Bundle
-	default:
-		return fmt.Errorf("please provide a cert to verify against via --certificate or a bundle via --bundle")
+		co.SCT = sct
 	}
-
-	// Performs all blob verification.
-	if err := verifyBlob(ctx, co, blobBytes, sig, cert, bundle); err != nil {
-		return err
+	// Set a cert chain if provided.
+	var chainPEM []byte
+	if c.CertChain != "" {
+		chain, err := loadCertChainFromFileOrURL(c.CertChain)
+		if err != nil {
+			return err
+		}
+		if chain != nil {
+			// Set the last one in the co.RootCerts. This is trusted, as its passed in
+			// via the CLI.
+			if co.RootCerts == nil {
+				co.RootCerts = x509.NewCertPool()
+			}
+			co.RootCerts.AddCert(chain[len(chain)-1])
+			// Use the rest as the cert chain in the signature object.
```

Comment:
nit: It's not the rest, it's the entire chain (intermediates and root) - this is intentional, correct?

Reply:
Yep it's intentional, I'll fix the comment. The cert chain in the signature object is taken "with a grain of salt" (omits the last cert)